debezium-connector-oracle/src/main/java/io/debezium/connector/oracle/logminer/LogMinerHistoryRecorder.java

```diff
@@ -0,0 +1,296 @@
+/*
+ * Copyright Debezium Authors.
+ *
+ * Licensed under the Apache Software License version 2.0, available at http://www.apache.org/licenses/LICENSE-2.0
+ */
+package io.debezium.connector.oracle.logminer;
+
+import java.math.BigDecimal;
+import java.sql.CallableStatement;
+import java.sql.PreparedStatement;
+import java.sql.SQLDataException;
+import java.sql.SQLException;
+import java.sql.Timestamp;
+import java.time.Duration;
+import java.time.Instant;
+import java.time.LocalDateTime;
+import java.util.concurrent.ArrayBlockingQueue;
+import java.util.concurrent.BlockingQueue;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.RejectedExecutionException;
+import java.util.concurrent.ThreadPoolExecutor;
+import java.util.concurrent.TimeUnit;
+import java.util.function.Supplier;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import io.debezium.DebeziumException;
+import io.debezium.annotation.NotThreadSafe;
+import io.debezium.connector.oracle.OracleConnection;
+import io.debezium.connector.oracle.OracleConnector;
+import io.debezium.jdbc.JdbcConfiguration;
+import io.debezium.util.Threads;
+
+/**
+ * This class appends captured incremental changes into log miner history table.
+ * Use cases:
+ * 1. Troubleshooting/viewing history for auditing or debugging purposes
+ * 2. Replaying recorded history
+ *    - for regression testing. This will not cover all application part, like parsing, because we record already parsed redoSql.
+ *      it could be beneficial for consumer testing for instance
+ *    - if Kafka is down (further code modification needed)
+ *    - for database migration. Do consistent dump using DPDUMP utility and replay the tail from this history table.
+ * todo ensure thread safeness for multi-threaded parser branch
+ * todo allow dynamic initialization after resolving DB problems
+ */
+@NotThreadSafe
+public class LogMinerHistoryRecorder implements HistoryRecorder, AutoCloseable {
+
+    private ExecutorService executor;
+    private Supplier<Integer> recordQueueCapacity;
+
+    private static int HISTORY_TABLE_SIZE_LIMIT = 1_000_000;
+
+    private volatile PreparedStatement psInsert;
+    private volatile CallableStatement bulkInsertCall;
+    private volatile CallableStatement truncateTempTable;
+
+    private OracleConnection recordHistoryConnection;
+    private final LogMinerMetrics metrics;
+    private JdbcConfiguration configuration;
+    private long historyRetention;
+    private int errorCounter = 0;
+
+    private final static Logger LOGGER = LoggerFactory.getLogger(LogMinerHistoryRecorder.class);
+
+    public LogMinerHistoryRecorder(LogMinerMetrics metrics, JdbcConfiguration configuration, long historyRetention) {
+        this.metrics = metrics;
+        this.configuration = configuration;
+        this.historyRetention = historyRetention;
+
+        try {
+            createExecutor();
```

Comment:
@AndreyIg Is there a reason why an executor, separate thread & connection necessary for this recorder logic?

Reply:
The main idea to keep this logic separate. If something happens, like tablespace cannot be extended, we stop recording with no affect on the main logic.
It also has separate queue.

Reply:
Sounds good.  Since I've removed `LogMinerHistoryRecorder` implementation from upstream Debezium, how you intend to manage the process is entirely now user-specific.  As we move forward, if there are additional hooks needed on the `HistoryRecorder` interface, let us know.